model/schema.go

```diff
@@ -0,0 +1,100 @@
+package model
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+// Schema is a JSON schema for a stream.
+type Schema struct {
+	// Type is the type of the schema, e.g. "object"
+	Type []string `json:"type"`
+	// HasAdditionalProperties indicates whether the schema allows additional properties
+	// not defined in the schema.
+	HasAdditionalProperties bool `json:"additionalProperties"`
+	// Properties is a map of property names to their schema.
+	Properties map[string]Property `json:"properties"`
+}
+
+// Property is a property in a JSON schema.
+type Property struct {
+	// Types is a list of types that this property can be, e.g. "string" or "integer".
+	Types []string `json:"type"`
+	// CustomFormat	is a custom format for this property, e.g. "date-time".
+	CustomFormat string `json:"format,omitempty"`
+	// For nested structures a property can have its own properties.
+	Properties map[string]Property `json:"properties,omitempty"`
+	// For array structures we define the type of the items in the array
+	Items *ArrayItem `json:"items,omitempty"`
+}
+
+// ArrayItem is the type and properties of an item in an array.
+type ArrayItem struct {
+	Type       string              `json:"type"`
+	Properties map[string]Property `json:"properties,omitempty"`
+}
+
+func (s Property) IsBoolean() bool {
+	return s.hasType("boolean")
+}
+
+func (s Property) IsNumber() bool {
+	return s.hasType("number")
+}
+
+func (s Property) IsInteger() bool {
+	return s.hasType("integer")
+}
+
+func (s Property) hasType(typeName string) bool {
+	for _, t := range s.Types {
+		if strings.EqualFold(t, typeName) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s Property) IsDateTime() bool {
+	return s.CustomFormat == "date-time"
+}
+
+// As a shortcut for outputting some content we can json back and forth to turn it into
+// the map[string]any that we need for the output
+// But if we're going to just turn this all into json why bother? Why not do this for everything?
+// Initial thoughts were that we want some control on the fields we output potentially - for example
+// ignoring deprecated fields.
+// We also might need to be cleverer when it comes to catalog config that enables / disables optional fields
+// in the output.
+func StructToMap(input interface{}) map[string]any {
```

Comment:
FYI there's https://pkg.go.dev/github.com/fatih/structs for this!

Reply:
Aaah perfect, thanks!